mbs_results/constrains.py

```diff
@@ -0,0 +1,170 @@
+    ----------
+    df : pd.DataFrame
+        Original dataframe to replace the values.
+    target : str
+        Column name for values to be replced.
+    a : int
+        Question_no to check.
+    compare : str
+        Logical operator to compare, accepted '>', '<','>=','<=','=='.
+    b : int
+        Question_no to check against.
+    """
+    # For improved perfomance
+    df.sort_index(inplace=True)
+
+    ops = {
+        ">": operator.gt,
+        "<": operator.lt,
+        ">=": operator.ge,
+        "<=": operator.le,
+        "==": operator.eq,
+    }
+
+    series_from_a = df.loc[a][target]
+
+    series_from_b = df.loc[b][target]
+
+    common_index = series_from_a.index.intersection(series_from_b.index)
+
+    # Has format (period,reference)
+    index_to_replace = series_from_a[
+        ops[compare](series_from_a[common_index], series_from_b[common_index])
+    ].index
+
+    if len(index_to_replace) > 0:
+        for date_ref_idx in index_to_replace.values:
+
+            # Has format (question,no,period,reference)
+            index_to_replace = (a,) + date_ref_idx
+            index_to_replace_with = (b,) + date_ref_idx
+            # Filter target based on the indices
+            df.loc[index_to_replace, target] = df.loc[index_to_replace_with, target]
+            df.loc[index_to_replace, "constrain_marker"] = f"{a} {compare} {b}"
+
+
+def sum_sub_df(df: pd.DataFrame, derive_from: List[int]) -> pd.DataFrame:
+    """
+    Returns the sums of a dataframe in which the first level index is in
+    derive_from. The sum is based on common indices. Columns must contain float
+    or int.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Dataframe to sum, first level index must contain values from derive_from
+
+    derive_from : List[int]
+        Values to take a subset of df.
+
+    Returns
+    -------
+    sums : pd.DataFrame
+        A dataframe with sums, constain marker, and columns from index which the
+        sum was based on.
+    """
+
+    sums = sum(
+        [df.loc[question_no] for question_no in derive_from if question_no in df.index]
+    )
+
+    return sums.assign(constrain_marker=f"sum{derive_from}").reset_index()
+
+
+def constrain(
+    df: pd.DataFrame,
+    period: str,
+    reference: str,
+    target: str,
+    target_imputed: str,
+    question_no: str,
+    spp_form_id: str,
+) -> pd.DataFrame:
+    """
+    Creates new rows with derived values based on form id and adds a relevant
+    marker to constain_marker column (is created if not existing).
+
+        For form id 13, question number 40 is created by summing 46,47.
+        For form id 14, question number 40 is created by summing 42,43.
+        For form id 15, question number 46 is created from 40.
+        For form id 16, question number 42 is created from 40.
+
+    In addition for all form types (when question number is available):
+
+        Replaces 49 with 40 when 49 > 40.
+        Replaces 90 with 40 when 49 >= 40.
```

Comment:
Double check second constraint is correct. I don't know anything about these, but it reads like it might be based on 90 and not 49 in the second case

Reply:
Good spot, it was a typo in the docs